### PR TITLE
Fix for clang-tidy 16

### PR DESCRIPTION
**Pull request by chaeyeunpark**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:** Some additional criteria in `clang-tidy-16` do not like our current codebase. This simple PR fixes it. Still, we do not update the `clang-tidy` version of the CI yet.

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane_lightning/src/tests/CMakeLists.txt [hunk 1]
```diff
@@ -19,6 +19,9 @@ FetchContent_Declare(
 
 FetchContent_MakeAvailable(Catch2)
 
+get_target_property(CATCH2_IID Catch2 INTERFACE_INCLUDE_DIRECTORIES)
+set_target_properties(Catch2 PROPERTIES INTERFACE_SYSTEM_INCLUDE_DIRECTORIES "${CATCH2_IID}")
```
Comment (chaeyeunpark) [inline pennylane_lightning/src/tests/CMakeLists.txt:23]: These lines ensure Catch2 code is excluded from `clang-tidy`.

File: pennylane_lightning/src/gates/DynamicDispatcher.hpp [hunk 2]
```diff
@@ -134,6 +134,7 @@ template <typename PrecisionT> class DynamicDispatcher {
         -> std::vector<Gates::KernelType> {
         std::vector<Gates::KernelType> kernels;
 
+        kernels.reserve(kernel_names_.size());
```
Comment (mlxd) [inline pennylane_lightning/src/gates/DynamicDispatcher.hpp:137]: CT16 complained about this?

File: pennylane_lightning/src/tests/CMakeLists.txt (hunk 1 above)
Comment (mlxd) [reply pennylane_lightning/src/tests/CMakeLists.txt:23]: So none of the test macros get tidied? Or our Catch2 tests are excluded?

So the trick is to get the included headers from the library, and reclassify them as system includes, so then CT ignores them, as they are treated as not part of the package?

**Review by mlxd — COMMENTED** [review]
Thanks @chaeyeunpark 
Just a few quick q's

File: pennylane_lightning/src/tests/CMakeLists.txt (hunk 1 above)
Comment (chaeyeunpark) [reply pennylane_lightning/src/tests/CMakeLists.txt:23]: We just want to exclude Catch2 header files from `clang-tidy` as `clang-tidy` does not like a few lines of the code here. The same technique is also used for Kokkos https://github.com/PennyLaneAI/pennylane-lightning/blob/736f7826772ca2182c7690c04af3454c72900037/cmake/process_options.cmake#L92-L95

File: pennylane_lightning/src/gates/DynamicDispatcher.hpp (hunk 2 above)
Comment (chaeyeunpark) [reply pennylane_lightning/src/gates/DynamicDispatcher.hpp:137]: Yeap. If the size of the vector is known prior, CT16 requires reserving the memory first as it is performant.

File: pennylane_lightning/src/tests/CMakeLists.txt (hunk 1 above)
Comment (mlxd) [reply pennylane_lightning/src/tests/CMakeLists.txt:23]: Got it, thanks.

**Review by mlxd — APPROVED** [review]

**Review by AmintorDusko — APPROVED** [review]
Nice! Thank you for that!

**Pull request merged**